webapp/tests/karma/unit/services/db-sync.js

```diff
@@ -236,13 +236,85 @@ describe('DBSync service', () => {
         expect(onUpdate.args[1][0]).to.deep.eq({ to: 'success', from: 'success' });
       });
     });
+
+    describe('retries with smaller batch size', () => {
+
+      let count;
+      let retries;
+      const recursiveOnTo = sinon.stub();
+
+      const replicationResultTo = () => {
+        if (count <= retries) {
+          count++;
+          return Q.reject({ code: 413 });
+        } else {
+          return Q.resolve();
+        }
+      };
+
+      beforeEach(() => {
+        count = 0;
+
+        isOnlineOnly.returns(false);
+        Auth.resolves();
+
+        recursiveOnTo.callsFake(() => {
+          const promise = replicationResultTo();
+          promise.on = recursiveOnTo;
+          return promise;
+        });
+
+        to.callsFake(() => {
+          let promise;
+          if (count < retries) {
+            // Too big - retry
+            promise = Q.reject({ code: 413 });
+          } else {
+            // small enough - complete
+            promise = Q.resolve();
+          }
+          promise.on = recursiveOnTo;
+          return promise;
+        });
+      });
+
+      it('if request too large', () => {
+        retries = 3;
+        return service.sync().then(() => {
+          expect(Auth.callCount).to.equal(1);
+          expect(from.callCount).to.equal(1);
+          expect(to.callCount).to.equal(3);
+          expect(to.args[0][1].batch_size).to.equal(100);
+          expect(to.args[1][1].batch_size).to.equal(50);
+          expect(to.args[2][1].batch_size).to.equal(25);
+        });
+      });
+
+      it('gives up once batch size is 1', () => {
+        retries = 100; // should not get this far...
+        return service.sync().then(() => {
+          expect(from.callCount).to.equal(1);
+          expect(to.callCount).to.equal(7);
+          expect(to.args[0][1].batch_size).to.equal(100);
+          expect(to.args[1][1].batch_size).to.equal(50);
+          expect(to.args[2][1].batch_size).to.equal(25);
+          expect(to.args[3][1].batch_size).to.equal(12);
+          expect(to.args[4][1].batch_size).to.equal(6);
+          expect(to.args[5][1].batch_size).to.equal(3);
+          expect(to.args[6][1].batch_size).to.equal(1);
+          console.log('finishing up');
+        });
+      });
+
+    });
+
   });
 
   describe('replicateTo filter', () => {
 
     let filterFunction;
 
-    before(() => {
+    beforeEach(() => {
```

Comment:
This had to change because otherwise the `beforeEach` in the `describe` section above ran after it and stuffed up the tests.